pkg/apis/core/validation/validation.go

```diff
@@ -7253,3 +7289,88 @@ func ValidatePodAffinityTermSelector(podAffinityTerm core.PodAffinityTerm, allow
 	return allErrs
 }
+
+func validateNodeSelectorMutation(fldPath *field.Path, oldNodeSelector, newNodeSelector map[string]string) field.ErrorList {
+	var allErrs field.ErrorList
+	// If old node selector was nil, anything can be set.
+	if oldNodeSelector == nil {
+		return allErrs
+	}
+	// Validate no existing node selectors were deleted or mutated.
+	for k, v1 := range oldNodeSelector {
+		if v2, ok := newNodeSelector[k]; !ok || v1 != v2 {
+			allErrs = append(allErrs, field.Invalid(fldPath, newNodeSelector, "only additions to spec.nodeSelector are allowed (no mutations or deletions)"))
+			return allErrs
+		}
+	}
+	return allErrs
+}
+
+func validateNodeAffinityMutation(nodeAffinityPath *field.Path, oldNodeAffinity, newNodeAffinity *core.NodeAffinity) field.ErrorList {
+	var allErrs field.ErrorList
+	// If old node affinity was nil, anything can be set.
+	if oldNodeAffinity == nil {
+		return allErrs
+	}
+
+	// RequiredDuringSchedulingIgnoredDuringExecution must not be nil for both old and new nodeAffinity
+	// in order to perform this validation.
+	if oldNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution == nil || newNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution == nil {
+		return allErrs
+	}
+
+	// NodeSelectorTerms are semantically OR-ed, so adding a new one would
+	// broaden the constraint rather than limit it further.
+	if len(oldNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms) != len(newNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms) {
+		return append(allErrs, field.Invalid(nodeAffinityPath.Child("requiredDuringSchedulingIgnoredDuringExecution").Child("nodeSelectorTerms"), newNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms, "no additions to non-empty NodeSelectorTerms list are allowed"))
+	}
+
+	// For requiredDuringSchedulingIgnoredDuringExecution, if old NodeSelectorTerms
+	// was empty, anything can be set. If non-empty, only additions of NodeSelectorRequirements
+	// to matchExpressions or fieldExpressions are allowed.
+	for _, oldTerm := range oldNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms {
+		found := false
+		for _, newTerm := range newNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms {
+			if validateNodeSelectorTermHasOnlyAdditions(oldTerm, newTerm) {
+				found = true
+				break
+			}
+		}
+		if !found {
+			allErrs = append(allErrs, field.Invalid(nodeAffinityPath.Child("requiredDuringSchedulingIgnoredDuringExecution").Child("nodeSelectorTerms"), oldTerm, "only additions are allowed (no mutations or deletions)"))
+			return allErrs
+		}
+	}
+	return allErrs
+}
+
+func validateNodeSelectorTermHasOnlyAdditions(oldTerm, newTerm core.NodeSelectorTerm) bool {
+	// Check MatchExpressions.
+	for _, oldExpr := range oldTerm.MatchExpressions {
+		found := false
+		for _, newExpr := range newTerm.MatchExpressions {
```

Comment:
this lets old expressions move location... is there a reason for this flexibility? I expected:

```go
if l := len(oldTerm.MatchExpressions); l > 0 {
  if len(newTerm.MatchExpressions) < l {
    return false
  }
  if !reflect.DeepEqual(newTerm.MatchExpressions[:l-1], oldTerm.MatchExpressions) {
    return false
  }
}
```